File: apps/api/lib/nxtvibes/chats/group_chat.ex
defmodule NxtVibes.Chats.GroupChat do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "group_chats" do
    # Real columns from Supabase — confirmed via information_schema
    field :group_name, :string
    field :group_description, :string
    field :group_icon, :string
    field :participants, {:array, :string}, default: []
    field :participant_details, :map, default: %{}
    field :created_by, :string
    field :member_count, :integer
    field :hidden, :boolean, default: false
    field :admins, {:array, :string}, default: []
    field :last_message, :map
    field :unread_count, :map, default: %{}
    field :deleted_by, {:array, :string}, default: []   # actual column: deleted_by (not deleted_for)
    field :cleared_at, :map, default: %{}
    field :typing, :map, default: %{}

    timestamps(type: :utc_datetime, inserted_at: :created_at, updated_at: :updated_at)
  end

  @doc false
  def changeset(group_chat, attrs) do
    group_chat
    |> cast(attrs, [
      :group_name,
      :group_description,
      :group_icon,
      :participants,
      :participant_details,
      :created_by,
      :member_count,
      :hidden,
      :admins,
      :last_message,
      :unread_count,
      :deleted_by,
      :cleared_at,
      :typing
    ])
    |> validate_required([:group_name, :participants])
  end
end

File: apps/mobile/src/database/models/GroupChat.ts
import { Model } from '@nozbe/watermelondb';
import { field, date, children, readonly } from '@nozbe/watermelondb/decorators';

export default class GroupChat extends Model {
  static table = 'group_chats';
  static associations = {
    messages: { type: 'has_many', foreignKey: 'chat_id' },
  } as const;

  @field('group_name') groupName!: string;
  @field('group_description') groupDescription?: string;
  @field('group_icon') groupIcon?: string;
  @field('participants') participantsRaw!: string; // JSON string
  @field('participant_details') participantDetailsRaw?: string; // JSON string
  @field('created_by') createdBy?: string;
  @field('member_count') memberCount?: number;
  @field('hidden') hidden!: boolean;
  @field('admins') adminsRaw?: string; // JSON string
  @field('last_message') lastMessage?: string;
  @field('unread_count') unreadCountRaw?: string; // JSON string
  @field('deleted_for') deletedForRaw?: string; // JSON string
  @field('cleared_at') clearedAtRaw?: string; // JSON string
  // typing is ephemeral — tracked via Phoenix Presence, not persisted in DB
  @field('muted_by') mutedByRaw?: string; // JSON string (array of user IDs)
  @field('pinned_by') pinnedByRaw?: string; // JSON string (array of user IDs)

  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;

  @children('messages') messages!: any;

  get participants(): string[] {
    try {
      return this.participantsRaw ? JSON.parse(this.participantsRaw) : [];
    } catch {
      return [];
    }
  }

  get unreadCount(): Record<string, number> {
    try {
      return this.unreadCountRaw ? JSON.parse(this.unreadCountRaw) : {};
    } catch {
      return {};
    }
  }
}
